dynamic_instrumentation/src/instrument_memory.rs

```diff
@@ -236,16 +236,62 @@ impl<'a, 'tcx: 'a> CollectFunctionInstrumentationPoints<'a, 'tcx> {
         find_instrumentation_def(self.tcx, self.runtime_crate_did, name)
     }
 
-    fn should_instrument(&self, local: Local, context: PlaceContext) -> bool {
-        let is_unsafe_ptr = self.body.local_decls[local].ty.is_unsafe_ptr();
-        is_unsafe_ptr && context.is_use()
+    fn is_shared_or_unsafe_ptr(&self, ty: &TyS<'tcx>) -> bool {
+        ty.is_unsafe_ptr() || (ty.is_region_ptr() && !ty.is_mutable_ptr())
     }
 
     fn func_hash(&self) -> mir_loc::DefPathHash {
         self.tcx.def_path_hash(self.body.source.def_id()).convert()
     }
 }
 
+fn has_outer_deref(p: &Place) -> bool {
+    matches!(
+        p.iter_projections().last(),
+        Some((_, ProjectionElem::Deref))
+    )
+}
+
+fn pop_last_projection<'tcx>(p: &Place<'tcx>, tcx: TyCtxt<'tcx>) -> Option<Place<'tcx>> {
+    if p.projection.is_empty() {
+        return None;
+    }
+    Some(Place {
+        local: p.local,
+        projection: tcx.intern_place_elems(&p.projection[..p.projection.len() - 1]),
+    })
+}
```

Comment:
I refactored this using `split_last`; see https://github.com/immunant/c2rust/pull/521/commits/b29e72608b2b26cceea797f963179e4cbf79cdf4.  Does that look fine?